Convert Watch Status and Action Status model classes to stateless functions and TS

cjcenizal opened the pull request:
# Summary

This change is intended to make the server-side Watch Status and Action Status code easier to understand and more robust through TS guarantees, in order to reduce the risk of regressions when we make changes to it this logic later.

## Original model

Originally this code followed an object-oriented paradigm and was composed of two classes: `ActionStatusModel` and `WatchStatusModel`. Each of these classes exposed two common methods:

* **Factory function:** `static fromUpstreamJson(json)` This was used to create instances of each class. Typically very little processing was done on the data passed in as the `json` argument. At most, timestamp strings were converted to Moment instances.
* **Deserialization getter:** `get downstreamJson()` This was called prior to returning a response to the Kibana client. It does more processing by calling specialized getter functions that would derive various state.

In addition to these methods, each class also exposed **specialized getter functions** for deriving various state.

This design created two problems:
* **Confusing data flow.** The factory function and deserialization getter offer little cohesion by name or signature, making their roles in the data flow unclear. The concept of "upstream" and "downstream" is confusing because these are all relative concepts in a system like a server-client relationship that maintains a bidirectional data flow. It's also confusing when there are three actors involved (ES, Kibana server, Kibana client).
* **Exposed implementation details.** The specialized getter functions are part of the public interfaces of these classes, but they are intended for internal consumption. This obscures their role and opens the door for misuse.

## Approach

I defined a boundary at the watch status and action status models. My changes are limited to this code, and I left other models as untouched as possible (BaseWatch, WatchHistoryItem).

I decided to solve the above problems by implementing two strategies:
* **Functional paradigm.** I've replaced the OO paradigm with a functional paradigm that uses stateless functions and plain objects instead of classes. This makes the data flow easier to follow by enabling clearer names. It also enables us to hide implementation details by removing functions from the public interface.
* **Move away from "upstream" and "downstream" concepts.** I've replaced these concepts with names that reflect the intended consumer (Kibana server, Kibana client, ES).

# For reviewers

Many of the new types reference those provided by the Elasticsearch JS client. See the [published types](https://github.com/elastic/elasticsearch-js/blob/main/src/api/types.ts) for cross-reference.

## Steps to test

Create a watch locally with this request and explore the UI.

```
PUT _watcher/watch/test
{
  "trigger": {
    "schedule": {
      "interval": "10m"
    }
  },
  "actions" : {
    "log" : { 
      "logging" : {
        "text" : "executed at {{ctx.execution_time}}" 
      }
    }
  }
}
```

# Next steps

There are two major next steps remaining once this PR is merged.

* Formalizing watchErrors.
* Implementing changes to states and state logic.

## Formalize watchErrors

The logic for generating watchErrors is challenging to follow and beyond the scope of this PR. If we continue with converting this codebase to TS, formalizing watchErrors will likely spiderweb throughout the codebase.

## Changes to states and state logic

After this is merged, I intend to make changes to the types of watch status and action status states that are in use, and to change how they're calculated. Here are the planned changes at a high-level.

See https://github.com/elastic/kibana/pull/138563.

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

alisonelizabeth commented on x-pack/plugins/watcher/common/constants/action_states.ts:
```diff
@@ -5,41 +5,31 @@
-import { i18n } from '@kbn/i18n';
-
+// These values are returned by the API and depended upon in both server-side
+// and client-side logic. TS will use these values to determine correctness.
```
What do you think about using snakecase for the values  (`ok`, `config_error`, `acknowledged`) since these will not surface in the UI?

cjcenizal replied:
Good idea. I'll implement this if I can't revert these changes, per my comment above.

alisonelizabeth commented on x-pack/plugins/watcher/common/constants/watch_states.ts:
```diff
@@ -5,26 +5,16 @@
-import { i18n } from '@kbn/i18n';
-
+// These values are returned by the API and depended upon in both server-side
+// and client-side logic. TS will use these values to determine correctness.
```
What do you think about using snakecase for the values  (`ok`, `config_error`, `disabled`) since these will not surface in the UI?

alisonelizabeth commented on x-pack/plugins/watcher/server/models/watch_status_model/watch_status_model.test.ts:
```diff
@@ -0,0 +1,134 @@
+
+describe('WatchStatusModel', () => {
+  describe('buildServerWatchStatusModel', () => {
+    it(`throws an error if no 'id' property in json`, () => {
```
Should we add a note that this test is temporary until all consumers are converted to TS?

alisonelizabeth commented on x-pack/plugins/watcher/server/models/watch_status_model/watch_status_model.test.ts:
```diff
@@ -0,0 +1,134 @@
+      }).toThrow(/must contain an id property/i);
+    });
+
+    it(`throws an error if no 'watchStatusJson' property in json`, () => {
```
Should we add a note that this test is temporary until all consumers are converted to TS?

alisonelizabeth commented on x-pack/plugins/watcher/common/types/status_types.ts:
```diff
@@ -0,0 +1,83 @@
+
+import { ACTION_STATES, WATCH_STATES, WATCH_STATE_COMMENTS } from '../constants';
+
+export interface ActionStatusUpstreamJson {
```
Based on your PR description, it seems like the intention was to move away from upstream/downstream concepts. Should this be renamed?

cjcenizal commented on x-pack/plugins/watcher/common/types/status_types.ts:
```diff
@@ -0,0 +1,83 @@
+
+import { ACTION_STATES, WATCH_STATES, WATCH_STATE_COMMENTS } from '../constants';
+
+export interface ActionStatusUpstreamJson {
```
Good point!

BaseWatch seems to create a WatchStatusModel by synthesizing values that it receives from the consumer:

https://github.com/elastic/kibana/blob/664ddc2a95ea797a454bf81b23fbac1a99c030a6/x-pack/plugins/watcher/server/models/watch/base_watch.js#L121-L145

WatchHistoryItem does something similar:

https://github.com/elastic/kibana/blob/664ddc2a95ea797a454bf81b23fbac1a99c030a6/x-pack/plugins/watcher/server/models/watch_history_item/watch_history_item.js#L25-L31

API route handlers are more straightforward, and generate models directly from the ES API response:

https://github.com/elastic/kibana/blob/664ddc2a95ea797a454bf81b23fbac1a99c030a6/x-pack/plugins/watcher/server/routes/api/watch/action/register_acknowledge_route.ts#L46-L51

https://github.com/elastic/kibana/blob/664ddc2a95ea797a454bf81b23fbac1a99c030a6/x-pack/plugins/watcher/server/routes/api/watch/register_activate_route.ts#L44-L49

https://github.com/elastic/kibana/blob/664ddc2a95ea797a454bf81b23fbac1a99c030a6/x-pack/plugins/watcher/server/routes/api/watch/register_deactivate_route.ts#L44-L49

I suppose regardless of the different ways the `ActionStatusUpstreamJson` data is generated, it's all going to come from ES. So what do you think of calling this `ActionStatusModelParams`? Or `ActionStatusModelEs`?

cjcenizal commented on x-pack/plugins/watcher/common/types/status_types.ts:
```diff
@@ -0,0 +1,83 @@
+}
+
+interface SerializedWatchStatus extends estypes.WatcherActivationStatus {
+  // actions: WatcherActions // Record<IndexName, WatcherActionStatus>
```
Will do

cjcenizal commented on x-pack/plugins/watcher/common/constants/action_states.ts:
```diff
@@ -5,41 +5,31 @@
+// These values are returned by the API and depended upon in both server-side
+// and client-side logic. TS will use these values to determine correctness.
+// If we were to translate them for the runtime, we would lose the correctness
+// guarantee. Therefore, we shouldn't translate them. Translation should occur
```
Yes, there are. I'm going to try reverting the changes to these files and see if I can defer them to when we make changes to the UI.

cjcenizal commented on x-pack/plugins/watcher/common/types/status_types.ts:
```diff
@@ -0,0 +1,83 @@
+export interface ServerActionStatusModel {
+  id: string;
+  actionStatusJson: estypes.WatcherActionStatus;
+  errors: any; // TODO
```
Yes, it will be addressed as part of the "Formalize watcheErrors" work. I'll update the comment to reflect that.

cjcenizal commented on x-pack/plugins/watcher/server/models/action_status_model/action_status_model.ts:
```diff
@@ -0,0 +1,80 @@
+  };
+};
+
+// Replaces downstreamJson getter method.
```
I'll remove these, thanks.

cjcenizal commented:
Thanks for your review @alisonelizabeth! I believe I addressed your comments. Could you please take another look? Could you please also compare the tests before and after these changes, and see if I might have missed anything? I forgot to mention in the PR description that I had to make significant changes to how they're implemented which introduces risk that we've lost coverage.

alisonelizabeth commented on x-pack/plugins/watcher/server/models/watch_status_model/watch_status_model.test.ts:
```diff
@@ -0,0 +1,142 @@
+  });
+
+  describe('buildClientWatchStatusModel', () => {
+    it('returns correct object for use by Kibana client', () => {
```
I know the comment and state fields are tested more heavily in `watch_status_model_utils.test`, but, for completeness here, should we also add expect statements for `serverWatchStatusModel.comment` and `serverWatchStatusModel.state`?

cjcenizal replied:
Done.

alisonelizabeth reviewed (approved):
Thanks for addressing my feedback! I gave the tests another look. I left one comment, but otherwise did not see any gaps in coverage. 

Side note - while running through the tests again, I got confused for a minute around the difference between `lastExecutionSuccessful` and `lastSuccessfulExecution`. At some point, I think it would be helpful to rename `lastExecutionSuccessful` to something like `isLastExecutionSuccessful` to more clearly indicate it's a boolean.
